lib/index.ts

```diff
@@ -1,28 +1,99 @@
-// convoluted example to show different compilation targets
-// are respected
-export default function(a: number, b: number): number {
-  const id = (x: any) => x
-  return id(a) + id(b)
-}
+import atxHeading from './atx-heading'
+import block from './block'
+import * as c from './characters'
+import paragraph from './paragraph'
 
-// examples below to check if the correct libs are loaded up
-// in your respective editors
+type ContextType = 'atxHeading' | 'paragraph' | 'block'
 
-// array.includes is from lib.es2016
-export function includes<T>(v: T, arr: T[]): boolean {
-  return arr.includes(v)
-}
+export class Tokenizer {
+  public data = ''
+  public line = 1
+  public column = 1
+  public virtualColumn = 1
+  public offset = 0
+  public tabSize = 2
 
-// padStart and padEnd are from lib.es2017
-export function padString(inputString: string, maxLength: number, fillWith?: string): string {
-  const strLen = inputString.length
+  public context: ContextType = 'block'
+  public stateHandlers: any
+  public contextInfo = {}
+  public state: string = ''
 
-  return inputString
-    .padStart(strLen + Math.floor(maxLength / 2), fillWith)
-    .padEnd(strLen + maxLength, fillWith)
-}
+  public block: any = block
+  public atxHeading: any = atxHeading
+  public paragraph: any = paragraph
+
+  constructor() {
+    this.switch('block')
+  }
+
+  public write(chunk: string) {
+    this.data += chunk
+
+    while (this.offset <= this.data.length) {
+      this.next()
+    }
+  }
+
+  public current() {
+    const length = this.data.length
+    const offset = this.offset
+
+    if (offset > length) {
+      throw new Error('Cannot read after end')
+    }
+
+    // Use `null` instead of NaN for EOF code.
+    if (length === offset) {
+      return null
+    }
+
+    // Use `�` instead of `\0` (https://spec.commonmark.org/0.28/#insecure-characters).
+    return this.data.charCodeAt(offset) || c.replacementCharacter
+  }
+
+  public now() {
+    const { line, column, offset } = this
+    return { line, column, offset }
+  }
+
+  public consume() {
+    const code = this.current()
+    const tabSize = this.tabSize
+
+    if (code === null || code === c.eof || code === c.lineFeed) {
+      this.line++
+      this.column = 0
+    } else if (code === c.tab) {
+      this.virtualColumn = Math.floor(this.virtualColumn / tabSize) * tabSize + tabSize
+    }
+
+    // TODO handling code === null
+    // tslint:disable-next-line:no-console
+    console.log('consume: %s', this.state, this.now(), [String.fromCharCode(code!)])
+
+    this.column++
+    this.offset++
+  }
+
+  public reconsume(state: string) {
```

Comment:
`state` can be a `keyof typeof ATXStateTypes | BlockStateTypes | ParagraphStateTypes` where each of these `StateTypes` is the exported `enum` from each of the respective files.